### PR TITLE
Fix training of decoder embeddings.

### PR DESCRIPTION
**Pull request by manipopopo**
When `loop_function` is provided, `rnn_decoder` and `attention_decoder` do not propagate gradients over the loop function. However,  when `feed_previous` for `embedding_rnn_decoder`, `embedding_tied_rnn_seq2seq` and `embedding_attention_decoder` is `True`, the output symbols are embedded by the loop function `extract_argmax_and_embed`. The embedding for output symbols will learning nothing if there is no gradient propagating over the loop function.

I think it will be better if the propagation of gradients is specified within the loop function itself instead of `rnn_decoder` and `attention_decoder`.

### CODE REVIEW

**Comment by vrv** [comment]
ping for @ebrevdo

**Comment by ebrevdo** [comment]
Lukasz knows this code better.

**Comment by lukaszkaiser** [comment]
I'm not sure it's always better to propagate down to the embeddings, but indeed - it would be great to at least have the option. Before accepting the code I think it's necessary to correct the embedding_seq2seq models that define loop functions extract_argmax_and_embed -- these functions will not work at all without stop_gradients (as there is no gradient for argmax). I hope our tests will catch this.

**Comment by manipopopo** [comment]
If the gradients don't propagate down to the embeddings, the input of the decoder RNN will be the random _initial values_ throughout the training. This might be weird, for example, when the embeddings for all words have the same initial values (word 'ils' corresponds to `[1, 0, 0, 0, 0]`, word 'personnes' corresponds to `[1, 0, 0, 0, 0]` too, ...).
Maybe it will be better if the embeddings could learn something during the training. And if it is not the case, maybe we could replace the return value of `extract_argmax_and_embed`, which is now `return embedding_ops.embedding_lookup(embedding, prev_symbol)`, with `return array_ops.stop_gradient(embedding_ops.embedding_lookup(embedding, prev_symbol))`.

I am not sure, but I think that gradients don't propagate down to the second parameter of `embedding_ops.embedding_lookup`. Could we replace `prev_symbol = array_ops.stop_gradient(math_ops.argmax(prev, 1))` in `extract_argmax_and_embed` with `prev_symbol = math_ops.argmax(prev, 1)`?

**Comment by lukaszkaiser** [comment]
In all cases, this patch needs a unit test.

**Comment by manipopopo** [comment]
I'll try to come up with a test.

**Comment by manipopopo** [comment]
When `feed_previous` is a boolean tensor, `embedding_rnn_seq2seq` will construct two `embedding_rnn_decoder` with `feed_previous=True` and `feed_previous=False`. During the course of training, gradients will propagate through the embeddings within the decoders when `feed_previous` is evaluated as false. 

However, when an user create `embedding_rnn_seq2seq` with `feed_previous=True`, there will be only one `embedding_rnn_decoder` with `feed_previous=True` in the model. It might be confusing that the embeddings within `embedding_rnn_decoder` learn nothing and remain as random initial values after training.

To control whether embeddings for symbols generated by the decoder itself will be updated during the course of training, a new parameter `update_embedding_for_previous` is added to `embedding_rnn_decoder` and `embedding_attention_decoder`. For example, the following two cases have the same effect:
- Train `embedding_rnn_seq2seq` with `feed_previous=True`, which contains a `embedding_rnn_decoder` with `feed_previous=True` and `update_embedding_for_previous=True`. The decoder is fed with "<Go>" and outputs "A, B, C".
- Train `embedding_rnn_seq2seq` with `feed_previous=False`. The decoder is fed with "<Go>, A, B".

A test unit is added to check this.

**Comment by lukaszkaiser** [comment]
This change looks very good to me, let's merge and submit. Thanks for correcting this error!

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
I see a conflict in this PRs future. A modified seq2seq is already on staging. I'd like to wait until we push this, and then resolve that conflict before we merge.

**Comment by manipopopo** [comment]
Hi @martinwicke, 
I've resolved some conflicts.

**Comment by vrv** [comment]
@tensorflow-jenkins: test this please

**Comment by manipopopo** [comment]
Hi @vrv ,
I added an empty line before the line.

**Comment by vrv** [comment]
@tensorflow-jenkins: test this please

**Comment by vrv** [comment]
@tensorflow-jenkins: test this please

**Comment by vrv** [comment]
Merged
